File: NeetCode 150/Linked List/143. Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return 
        
        # Step 1: Find middle using slow and fast pointers
        slow=head
        fast=head
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next
        
        # Step 2: Reverse the second half
        prev=None
        curr=slow.next #div is always like [1,2,3] [1,2]
        slow.next=None #imp # Split the list into two halves
        while curr:
            next_node=curr.next
            curr.next=prev
            prev=curr
            curr=next_node
        
         # Step 3: Merge two halves
        first=head
        second=prev
        while second: #since the div is always like [1,2,3] [1,2]
            tmp1,tmp2=first.next,second.next
            first.next = second
            second.next=tmp1
            first,second=tmp1,tmp2
        

File: NeedCode 150 (JAVA)/LinkedList/143. Reorder List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {

        if(head.next == null || head.next.next == null) return;

        //find middle
        ListNode slow = head, fast = head;
        while(fast.next !=null && fast.next.next !=null){
            slow = slow.next;
            fast = fast.next.next;
        }

        //reverse second
        ListNode second = reverse(slow.next);
        slow.next = null;

        //Merge
        ListNode first = head;
        while(second != null){
            ListNode temp1 = first.next;
            ListNode temp2 = second.next;

            first.next = second;
            second.next = temp1;

            first = temp1;
            second = temp2;
        }
        
    }

    private ListNode reverse(ListNode head){
        ListNode prev = null, curr = head;
        while(curr!=null){
            ListNode next = curr.next;
            curr.next = prev;
            prev  = curr;
            curr = next;
        }
        return prev;
    }
}
